File: integrated-project/mindboost-symfony/src/Service/UserProgressService.php
<?php

namespace App\Service;

use App\Entity\Score;
use App\Entity\SpecificScore;
use App\Entity\GeneralTest;
use App\Entity\SpecificTest;
use Doctrine\ORM\EntityManagerInterface;

class UserProgressService
{
    public function __construct(private EntityManagerInterface $entityManager)
    {
    }

    public function getUserHistory(int $userId): array
    {
        $generalScores = $this->entityManager->getRepository(Score::class)->findBy(
            ['userId' => $userId],
            ['id' => 'DESC']
        );

        $specificScores = $this->entityManager->getRepository(SpecificScore::class)->findBy(
            ['userId' => $userId],
            ['passedAt' => 'DESC']
        );

        $generalTests = $this->indexById($this->entityManager->getRepository(GeneralTest::class)->findAll());
        $specificTests = $this->indexById($this->entityManager->getRepository(SpecificTest::class)->findAll());

        $history = [];

        foreach ($generalScores as $score) {
            $history[] = [
                'type' => 'general',
                'title' => $generalTests[$score->getGeneralTestId()]?->getTitle() ?? 'Test général',
                'score' => $score->getTotalScore(),
                'percentage' => $score->getPercentage(),
                'category' => '-',
                'level' => '-',
                'date' => null,
            ];
        }

        foreach ($specificScores as $score) {
            $history[] = [
                'type' => 'specific',
                'title' => $specificTests[$score->getSpecificTestId()]?->getTitle() ?? 'Test spécifique',
                'score' => $score->getTotalScore(),
                'percentage' => $score->getPercentage(),
                'category' => $score->getCategory(),
                'level' => $score->getLevel(),
                'date' => $score->getPassedAt(),
            ];
        }

        usort($history, function (array $a, array $b) {
            $aDate = $a['date'] instanceof \DateTimeInterface ? $a['date']->getTimestamp() : 0;
            $bDate = $b['date'] instanceof \DateTimeInterface ? $b['date']->getTimestamp() : 0;

            return $bDate <=> $aDate;
        });

        return $history;
    }

    public function getEvolution(int $userId): array
    {
        $specificScores = $this->entityManager->getRepository(SpecificScore::class)->findBy(
            ['userId' => $userId],
            ['passedAt' => 'ASC']
        );

        $evolution = [];
        $previous = null;

        foreach ($specificScores as $score) {
            $trend = 'stable';
            $diff = 0;

            if ($previous !== null) {
                $diff = $score->getPercentage() - $previous->getPercentage();

                if ($diff > 0) {
                    $trend = 'amélioration';
                } elseif ($diff < 0) {
                    $trend = 'aggravation';
                }
            }

            $evolution[] = [
                'week' => $score->getWeekNumber(),
                'category' => $score->getCategory(),
                'level' => $score->getLevel(),
                'percentage' => $score->getPercentage(),
                'date' => $score->getPassedAt(),
                'difference' => $diff,
                'trend' => $trend,
            ];

            $previous = $score;
        }

        return $evolution;
    }

    private function indexById(array $entities): array
    {
        $indexed = [];

        foreach ($entities as $entity) {
            if (method_exists($entity, 'getId')) {
                $indexed[$entity->getId()] = $entity;
            }
        }

        return $indexed;
    }
}

File: integrated-project/mindboost-javafx/src/main/java/com/mindboost/app/service/UserProgressService.java
package com.mindboost.app.service;

import com.mindboost.app.config.DatabaseManager;
import com.mindboost.app.model.GeneralTest;
import com.mindboost.app.model.Score;
import com.mindboost.app.model.SpecificScore;
import com.mindboost.app.model.SpecificTest;
import jakarta.persistence.EntityManager;

import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.Comparator;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class UserProgressService {
    private final DatabaseManager databaseManager;

    public UserProgressService(DatabaseManager databaseManager) {
        this.databaseManager = databaseManager;
    }

    public List<HistoryItem> getUserHistory(int userId) {
        try (EntityManager em = databaseManager.createEntityManager()) {
            List<Score> generalScores = em.createQuery(
                "select s from Score s where s.userId = :userId order by s.id desc", Score.class)
                .setParameter("userId", userId)
                .getResultList();

            List<SpecificScore> specificScores = em.createQuery(
                "select s from SpecificScore s where s.userId = :userId order by s.passedAt desc", SpecificScore.class)
                .setParameter("userId", userId)
                .getResultList();

            Map<Integer, GeneralTest> generalTests = indexById(em.createQuery("from GeneralTest", GeneralTest.class).getResultList());
            Map<Integer, SpecificTest> specificTests = indexById(em.createQuery("from SpecificTest", SpecificTest.class).getResultList());

            List<HistoryItem> history = new ArrayList<>();

            for (Score score : generalScores) {
                GeneralTest test = generalTests.get(score.getGeneralTestId());
                history.add(new HistoryItem(
                    "general",
                    test == null ? "Test général" : test.getTitle(),
                    score.getTotalScore(),
                    score.getPercentage(),
                    "-",
                    "-",
                    null
                ));
            }

            for (SpecificScore score : specificScores) {
                SpecificTest test = specificTests.get(score.getSpecificTestId());
                history.add(new HistoryItem(
                    "specific",
                    test == null ? "Test spécifique" : test.getTitle(),
                    score.getTotalScore(),
                    score.getPercentage(),
                    score.getCategory(),
                    score.getLevel(),
                    score.getPassedAt()
                ));
            }

            history.sort(Comparator.comparing(HistoryItem::dateSafe).reversed());
            return history;
        }
    }

    public List<EvolutionItem> getEvolution(int userId) {
        try (EntityManager em = databaseManager.createEntityManager()) {
            List<SpecificScore> specificScores = em.createQuery(
                "select s from SpecificScore s where s.userId = :userId order by s.passedAt asc", SpecificScore.class)
                .setParameter("userId", userId)
                .getResultList();

            List<EvolutionItem> evolution = new ArrayList<>();
            SpecificScore previous = null;

            for (SpecificScore score : specificScores) {
                int diff = 0;
                String trend = "stable";
                if (previous != null) {
                    diff = score.getPercentage() - previous.getPercentage();
                    if (diff > 0) {
                        trend = "amélioration";
                    } else if (diff < 0) {
                        trend = "aggravation";
                    }
                }

                evolution.add(new EvolutionItem(
                    score.getWeekNumber(),
                    score.getCategory(),
                    score.getLevel(),
                    score.getPercentage(),
                    score.getPassedAt(),
                    diff,
                    trend
                ));
                previous = score;
            }

            return evolution;
        }
    }

    private <T> Map<Integer, T> indexById(List<T> entities) {
        Map<Integer, T> indexed = new HashMap<>();
        for (T entity : entities) {
            try {
                Integer id = (Integer) entity.getClass().getMethod("getId").invoke(entity);
                if (id != null) {
                    indexed.put(id, entity);
                }
            } catch (Exception ignored) {
            }
        }
        return indexed;
    }

    public record HistoryItem(String type, String title, Integer score, Integer percentage, String category, String level, LocalDateTime date) {
        public LocalDateTime dateSafe() {
            return date == null ? LocalDateTime.MIN : date;
        }
    }

    public record EvolutionItem(Integer week, String category, String level, Integer percentage, LocalDateTime date, Integer difference, String trend) {}
}
